### ISSUE TITLE
Support annotations on annotation assertions with templates

### ISSUE THREAD

**Issue by rctauber** [ontodev/robot#226]
Many ontologies use `database_cross_reference` as an annotation on a definition to cite sources. Both ECO and DO rely heavily on templates to create new terms, but we have to do this part manually.

I would like to look in to adding a new type of template string to support this. Not sure the best way to implement it on the template, but I was thinking something like:
```
A rdfs:comment [A rdfs:comment]
This is a comment [This is a comment on the comment]
```
My other thought was to do it as the column directly to the right of what you want to annotate with `AA` to specify that it's an annotation on that:
```
A rdfs:comment       AA rdfs:comment
This is a comment    This is a comment on the comment
```
I'd appreciate any input! I'm going to start working on the operation first.

**Comment by rctauber** [comment]
Got it working with the first method - that seems to be the easiest way. I changed the `get*Annotation` methods to return axioms instead of annotations. I pushed the changes to a new branch on my fork.

I did test it out briefly with this csv:
```
id,comment,other comment
ID,A rdfs:comment [rdfs:comment],A:rdfs:comment
ID:1,Comment 1 [Comment 2],Comment 3
```

And it seemed to work fine (yay!)
```
<Class rdf:about="ID:1">
    <rdfs:comment rdf:datatype="http://www.w3.org/2001/XMLSchema#string">Comment 1</rdfs:comment>
    <rdfs:comment rdf:datatype="http://www.w3.org/2001/XMLSchema#string">Comment 3</rdfs:comment>
</Class>
<Axiom>
    <annotatedSource rdf:resource="ID:1"/>
    <annotatedProperty rdf:resource="http://www.w3.org/2000/01/rdf-schema#comment"/>
    <annotatedTarget rdf:datatype="http://www.w3.org/2001/XMLSchema#string">Comment 1</annotatedTarget>
    <rdfs:comment rdf:datatype="http://www.w3.org/2001/XMLSchema#string">Comment 2</rdfs:comment>
</Axiom>
```
I'm going to play around with it more tomorrow, since I have a handful of new DO terms to add.

**Comment by jamesaoverton** [comment]
That output looks right.

I prefer the second option above: using separate columns. I think it's more natural, won't get mixed up with content of the annotation (e.g. square brackets in the content), and should allow for multiple annotations.

**Comment by rctauber** [comment]
I'll try and sort that out. That makes things a little more complicated (which I didn't think about before I dove into the code)... it could be done with another loop though.

**Comment by jamesaoverton** [comment]
Another nice-to-have (but not required) to consider is a third-level: annotation-on-annotation-on-annotation. I've occasionally seen this in the wild, but never more nesting than that. Implementation requires keeping a stack of annotations. HOWL/Knotation implements this kind of thing.

**Comment by rctauber** [comment]
Is this supported in Protege? I started to code it out and ran into the issue that there's no method to append assertion axioms onto an annotation, which creating an assertion axiom is the only way OWLAPI provides to make a nested axiom.

**Comment by rctauber** [comment]
Never mind, I found a work around!

**Comment by rctauber** [comment]
Added support for up to 3 levels of nesting. If there is an `AAA`, it requires an `AA` to the left of it, or an exception is thrown. Similarly, if there isn't an annotation next to the `AA` template, an exception is thrown.

| id | level 1 | level 2 | level 3 | level 3 | level 2 | level 1 |
| --- | --- | --- | --- | --- | --- | --- |
| ID | A rdfs:comment | AA rdfs:comment | AAA rdfs:comment | AAA rdfs:comment | AA rdfs:comment | A rdfs:comment |
| ID:1 | Comment 1 | Comment 2 | Comment 3 | Comment 4 | Comment 5 | Comment 6 |

<img width="308" alt="screen shot 2018-01-15 at 10 21 28 am" src="https://user-images.githubusercontent.com/16750150/34949429-161dc054-f9de-11e7-969f-ff3f770c95f5.png">

**Comment by jamesaoverton** [comment]
Looks good! Make a PR when you're ready.

**Comment by cmungall** [comment]
Minor terminological quibble:

the general case here is annotations on any kind of axiom, either Annotation Assertion Axioms (e.g. `C rdfs:comment foo`) or Logical Axioms (e.g. `C SubClassOf D`). The latter case is common in ontologies like Uberon and will likely become more common across multiple ontologies.

I generally avoid referring to "annotation assertion axioms" as "annotations", because these are different things in the OWL datamodel (although admittedly frame-syntaxes like Manchester blur this).

My preference would be to denote both of these as having __one__ level of nesting of annotations:

```
SubClassOf(Annotation(source "PMID:1") C D)
AnnotationAssertion(Annotation(source "PMID:1") rdfs:comment C "foo")
```

Structurally there are analogous - one axiom plus one annotation on the axiom. the latter happens to be an annotation assertion axiom, but ideally everything should be generic here.

Of course, the structural depth of the OWL functional syntax tree is 2 in both cases, but that would be a different measure than nesting of *annotations*. 

You win the OWLAPI guru award for getting support for 2 levels of annotations in there (or 3 by your scheme). Having >1 makes my head hurt. I can't really foresee a case where this is ever required. I've only seen this in the wild once, in ORDO, and I believe it was due to an error in OWL generation. Having up to 1 makes a lot of sense, it's a fairly standard data model: a statement and metadata about the statement. Lots of things like wikidata follow this (though they don't use the same reification structure, the model is the same). Now I've said all that I'm sure someone will prove me wrong...

I'm just being fussy about language here. I don't think any changes are required here. The scheme could be the same with the first `A` denoting "axioms" and the succeeding ones denoting annotations. I just want to make it clear in the documentation what's going on as it's easy for users to get confused by this stuff and it's much easier to set up conventions early on than fix people's mistakes after the fact...

**Comment by rctauber** [comment]
Ah, I didn't think about annotations on logical axioms. I'm going to add those in now.

It could also be an 'N' for annotation, like so:
```
A rdfs:comment,N rdfs:comment
Comment,Annotation comment
```
And...
```
C %,N rdfs:comment
CURIE,Annotation comment
```

And if you really want to: `A ...,N ..., NN ...` (it makes my head hurt too :smile:)

**Comment by jamesaoverton** [comment]
I dunno. In HOWL/Knotation I use `>`, `>>`.

**Comment by rctauber** [comment]
I would be easiest to do something like:
`C>` or `CN` for a logical expression annotation
`A>` or `AN` for an annotation annotation (`A>>` or `ANN` too)

Either way is fine by me.

**Comment by jamesaoverton** [comment]
I was thinking:

| id | level 1 | level 2 | level 3 | level 3 | level 2 | level 1 |
| --- | --- | --- | --- | --- | --- | --- |
| ID | A rdfs:comment | >A rdfs:comment | >>A rdfs:comment | >>A rdfs:comment | >A rdfs:comment | A rdfs:comment |
| ID:1 | Comment 1 | Comment 2 | Comment 3 | Comment 4 | Comment 5 | Comment 6 |

**Comment by rctauber** [comment]
Oh gotcha - that makes sense. I'll change it to that unless we come up with something else. 👍

**Comment by rctauber** [comment]
Closed by 19d4216
